Src/WindowsServer/WindowsServer.Shared/AzureWebAppRoleEnvironmentTelemetryInitializer.cs

```diff
@@ -62,7 +83,14 @@ private string GetRoleName()
 
         private string GetNodeName()
         {
-            return Environment.GetEnvironmentVariable(WebAppHostNameEnvironmentVariable) ?? string.Empty;
+            string nodeName = string.Empty;
+            AppServiceEnvVarMonitor.Instance.GetCurrentEnvironmentVariableValue(this.WebAppHostNameEnvironmentVariable, ref nodeName);
```

Comment:
Is this the quickest read possible out of ConcurrentDictionary (I see that internally it uses `GetOrAdd()`) - is there anything that reads faster than that for ConsurrentDictionary with no locks whatsoever?
We are OK if several items are still having old data, but having all ongoing requests/dependencies and other telemetry items to sync on read of that dictionary may lead to a bottleneck here.

Reply:
Good catch to call this out. However, I think I can put any fears to rest. 

Notice that by the time this call is being made, the cached value from the environment has already been stored for this update cycle, the key is already in place. This all works together to ensure the `GetOrAdd` call is simply a `TryGetValue` call ([that has no lock on it as you can see here](https://github.com/dotnet/corefx/blob/e0802d8b6010d4bb4538b15f228eee45378c4cc4/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L496)). 

However, in the scenario you suggest (in some future class that we decide to tax the `EnvironmentVariableMonitor` class as much as possible), it absolutely could be possible to create a nasty bottleneck.

Even then, as far as I am aware, GetOrAdd is the *best, fastest* way to deal with updates of this nature. 

`ConcurrentDictionary` actually does the only thing I can think of that *might* be faster, [breaking the problem down into a `TryGetValue` and then using a call to `TryUpdateValue` *only when things have changed*.](https://github.com/dotnet/corefx/blob/e0802d8b6010d4bb4538b15f228eee45378c4cc4/src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs#L1030)

In the scenario that you site though, where everything is all being updated at once, we still have a bottleneck (albeit one that is handled very nicely by `ConcurrentDictionary` without us having to write and maintain any complex code).

Reply:
I probably could have made this shorter:

> ... We are OK if several items are still having old data...

This is exactly the path that we follow today. 
1. Update all environment variables in a loop, detecting any change. 
2. If we detected a change, notify subscribers to collect their new cached values, ensuring only `TryGetValue` operations on *already tracked* items.

But thanks for making me read into how `ConcurrentDictionary` works under the hood. I do feel better about using it now. I was worried it was locking things even in the case of reads!

Reply:
Maybe I was diving too deep.

In the 'happy path', we always read from the local variables:

>From AzureWebAppRoleEnvironmentTelemetryInitializer.cs, line 45

    public void Initialize(ITelemetry telemetry)
    {
        if (this.updateEnvVars) // <= this is set in the update callback when change is detected
        {
            this.roleName = this.GetRoleName();
            this.nodeName = this.GetNodeName();
            this.updateEnvVars = false; // <= this ensures we don't check again until another change
        }
    ...

*Only when a change in the environment is detected* do we ever go and get the information from the monitor.